kotlin/authenticate-with-oauth/src/main/java/com/esri/arcgisruntime/sample/authenticatewithoauth/MainActivity.kt

```diff
@@ -0,0 +1,268 @@
+      0
+    )
+
+    // create an Intent to attempt to handle the authorization URL
+    with(Intent(Intent.ACTION_VIEW, Uri.parse(authorizationUrl))) {
+      this.resolveActivity(packageManager)?.let {
+        // this identifier ensures that the browser will attempt to reuse the same window each time the application
+        // launches the browser with the same identifier, which in this case is static and will always only be our
+        // application ID
+        this.putExtra(Browser.EXTRA_APPLICATION_ID, BuildConfig.APPLICATION_ID)
+        startActivity(this)
+        return
+      }
+    }
+
+    // user doesn't have a browser available to handle the Intent so we use a WebView to handle OAuth. WebView methods
+    // must be called on UI thread
+    runOnUiThread {
+      setupWebView()
+      webView.loadUrl(authorizationUrl)
+    }
+  }
+
+  /**
+   * Sets up the WebView.
+   *
+   * Please note that we have chosen to include use of the WebView as a fallback mechanism if the user does not have a
+   * web browser installed, only as a matter of course to allow you to use this sample in all circumstances. We do not
+   * recommend using WebView as part of your OAuth flow as per Google's recommendations:
+   * https://developers.googleblog.com/2016/08/modernizing-oauth-interactions-in-native-apps.html
+   */
+  private fun setupWebView() {
+    // setup a WebViewClient to override handling of custom scheme and host for Intent
+    webView.webViewClient = object : WebViewClient() {
+      override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
+        Uri.parse(url)?.let {
+          if (it.scheme == getString(R.string.oauth_redirect_scheme) && it.host == getString(R.string.oauth_redirect_host)) {
+            startActivity(generateAuthIntent(it))
+            return true
+          }
+        }
+        return super.shouldOverrideUrlLoading(view, url)
+      }
+
+      @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
+      override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
+        request?.url?.let { url ->
+          if (url.scheme == getString(R.string.oauth_redirect_scheme) && url.host == getString(R.string.oauth_redirect_host)) {
+            startActivity(generateAuthIntent(url))
+            return true
+          }
+        }
+        return super.shouldOverrideUrlLoading(view, request)
+      }
+    }
+
+    // enabled to allow javascript to run on auth webpage.
+    webView.settings.javaScriptEnabled = true
+    webView.visibility = View.VISIBLE
+  }
+
+  /**
+   * Generate the Intent that launches the MainActivity with the new auth code
+   *
+   * @param uri instance of Uri generated from URL that OAuth webpage redirects to after successful login
+   */
+  private fun generateAuthIntent(uri: Uri): Intent {
+    return Intent().also {
+      it.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+      it.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
```

Comment:
Can you add some comments here explaining what these flags are for? I assume they prevent the MainActivity from being destroyed?

Reply:
They're the flags for launching the browser `Intent`. It's `singleTop` on the `Activity` in the manifest that prevents the `Activity` being destroyed.